includes/admin/class-sensei-onboarding-api.php

```diff
@@ -0,0 +1,224 @@
+<?php
+/**
+ * Onboarding REST API.
+ *
+ * @package Sensei\Onboarding
+ * @since   3.1.0
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Sensei Onboarding REST API endpoints.
+ *
+ * @package Sensei
+ * @author  Automattic
+ * @since   3.1.0
+ */
+class Sensei_Onboarding_API {
```

Comment:
I mentioned this briefly in another PR, but we should probably use the pattern of REST API Controllers. See the REST API in WCPC for inspiration. For a more fully-featured example (but also more complex than we need right now), see the [WooCommerce REST API](https://github.com/woocommerce/woocommerce-rest-api).

Here is the doc for [Controller Classes](https://developer.wordpress.org/rest-api/extending-the-rest-api/controller-classes/).

Might even be best to have a controller for each page? We could maybe have a superclass that they all inherit from to share common code?

We should also use the "schema" for our endpoints as well, which would be easier to do with separate controllers.

Reply:
We are, and this is the REST API controller class. It registers the endpoints with a schema, and the callbacks mostly parse or format the data and call actual 'business logic' methods on core classes. 

Using the `WP_REST_Controller` or the specific pattern it provides is not a good fit here: we are not providing a REST API for CRUD operations on an entity, which that is geared towards. 

I considered splitting it up some more, but we have six endpoints each with about 2 lines of code, and no current plans for them to get significantly bigger, so this looks like a comfortable unit - separate files add some overhead that doesn't really seem to be worth it here.

Reply:
> this is the REST API controller class

Gotcha.

> Using the `WP_REST_Controller` or the specific pattern it provides is not a good fit here: we are not providing a REST API for CRUD operations on an entity, which that is geared towards.

Hmm. Well, there are parts of it that we can leave out, and certainly we don't want to do the `WP_REST_Posts_Controller` thing for this. But I think it's good to stick to known conventions, and architect our API in a standard way relative to other WP REST API's. I think there are a few things we can tweak make it more conventional:

**Filename/path:** Let's put it in `includes/rest-api` and name it consistently. I'm looking at [this controller](https://github.com/Automattic/sensei/blob/master/includes/rest-api/class-sensei-rest-api-messages-controller.php), and wondering if `Sensei_REST_API_Onboarding_Controller` might be clearer? (Unfortunately we aren't using PHP namespaces in Sensei yet, so awkward class names will be with us for a while).

**API Namespace:** It is conventional to specify the `namespace` near the top to make it clear. See this [example in WCPC](https://github.com/Automattic/sensei-wc-paid-courses/blob/master/includes/rest-api/controllers/class-course-products.php) and also this one in the [controller class documentation](https://developer.wordpress.org/rest-api/extending-the-rest-api/controller-classes/). Currently the namespace seems a bit buried in the code.

**Schema:** I do not see the schema declaration on the endpoint registration. Please see the [WCPC example](https://github.com/Automattic/sensei-wc-paid-courses/blob/master/includes/rest-api/controllers/class-course-products.php) and the [controller doc](https://developer.wordpress.org/rest-api/extending-the-rest-api/controller-classes/) for examples on how the schema is declared. Also see [this doc](https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/) for more background on the schema.

**Helper:** I see what you're doing with the helper method, but I think I would prefer to steer away from it if we could. It adds a small amount of convenience but IMO reduces clarity enough that it's not worth it. I personally think it would be better to call `register_rest_route` inline for each route. We could always put common args in variables that we then pull into each array, to improve maintainability.